File: backend/app/schemas/ota.py
"""OTA 升级 Schema — Story 15.5"""
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field


# --- Firmware ---
class FirmwareCreate(BaseModel):
    version: str = Field(..., max_length=50, description="版本号 (semver)")
    filename: str = Field(..., max_length=200, description="文件名")
    file_size: int = Field(..., gt=0, description="文件大小(字节)")
    checksum_sha256: str = Field(..., min_length=64, max_length=64, description="SHA-256 校验和")
    download_url: str = Field(..., max_length=500, description="下载地址")
    release_notes: Optional[str] = Field(None, max_length=2000, description="更新说明")
    min_version: Optional[str] = Field(None, max_length=50, description="最低兼容版本")


class FirmwareResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    version: str
    filename: str
    file_size: int
    checksum_sha256: str
    download_url: str
    release_notes: Optional[str] = None
    min_version: Optional[str] = None
    is_active: bool = True
    created_at: Optional[datetime] = None


# --- OTA Task ---
class OtaTaskCreate(BaseModel):
    firmware_id: int = Field(..., description="目标固件包 ID")
    gateway_ids: list[int] = Field(..., min_length=1, description="目标网关 ID 列表")
    strategy: str = Field("immediate", pattern="^(immediate|batch|canary)$", description="策略")
    batch_size: int = Field(0, ge=0, description="分批大小(0=全部)")
    batch_interval: int = Field(300, ge=0, le=3600, description="批次间隔(秒)")
    canary_percent: int = Field(10, ge=1, le=50, description="灰度百分比")


class OtaGatewayStatus(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    gateway_id: str
    batch_index: int = 0
    status: str = "pending"
    old_version: Optional[str] = None
    progress: int = 0
    error_message: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None


class OtaTaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    task_id: str
    firmware_id: int
    target_version: str
    strategy: str
    batch_size: int
    batch_interval: int
    canary_percent: int
    status: str
    total_gateways: int
    success_count: int
    fail_count: int
    created_by: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class OtaTaskDetailResponse(OtaTaskResponse):
    """任务详情（含各网关状态）"""
    gateways: list[OtaGatewayStatus] = []

File: frontend/src/api/modules/ota.ts
/**
 * OTA 升级管理 API — Story 15.5
 */
import request from '@/utils/request'
import type { PageParams } from './types'

// ==================== 类型定义 ====================

/** 固件包 */
export interface Firmware {
  id: number
  version: string
  filename: string
  file_size: number
  checksum_sha256: string
  download_url: string
  release_notes?: string
  min_version?: string
  is_active: boolean
  created_at?: string
}

/** 固件包创建参数 */
export interface FirmwareCreate {
  version: string
  filename: string
  file_size: number
  checksum_sha256: string
  download_url: string
  release_notes?: string
  min_version?: string
}

/** OTA 任务策略 */
export type OtaStrategy = 'immediate' | 'batch' | 'canary'

/** OTA 任务创建参数 */
export interface OtaTaskCreate {
  firmware_id: number
  gateway_ids: number[]
  strategy: OtaStrategy
  batch_size?: number
  batch_interval?: number
  canary_percent?: number
}

/** OTA 任务 */
export interface OtaTask {
  id: number
  task_id: string
  firmware_id: number
  target_version: string
  strategy: string
  batch_size: number
  batch_interval: number
  canary_percent: number
  status: string
  total_gateways: number
  success_count: number
  fail_count: number
  created_by?: string
  created_at?: string
  updated_at?: string
}

/** OTA 网关状态 */
export interface OtaGatewayStatus {
  gateway_id: string
  batch_index: number
  status: string
  old_version?: string
  progress: number
  error_message?: string
  started_at?: string
  completed_at?: string
}

/** OTA 任务详情 */
export interface OtaTaskDetail extends OtaTask {
  gateways: OtaGatewayStatus[]
}

// ==================== 固件包 API ====================

/** 获取固件包列表 */
export function getFirmwareList(params?: { is_active?: boolean }) {
  return request.get<Firmware[]>('/v1/ota/firmware', { params })
}

/** 注册固件包 */
export function createFirmware(data: FirmwareCreate) {
  return request.post<Firmware>('/v1/ota/firmware', data)
}

/** 删除固件包 */
export function deleteFirmware(id: number) {
  return request.delete(`/v1/ota/firmware/${id}`)
}

// ==================== OTA 任务 API ====================

/** 获取任务列表 */
export function getOtaTasks(params?: PageParams & { status?: string }) {
  return request.get<{ total: number; items: OtaTask[] }>('/v1/ota/tasks', { params })
}

/** 获取任务详情 */
export function getOtaTaskDetail(taskId: string) {
  return request.get<OtaTaskDetail>(`/v1/ota/tasks/${taskId}`)
}

/** 创建升级任务 */
export function createOtaTask(data: OtaTaskCreate) {
  return request.post<OtaTask>('/v1/ota/tasks', data)
}

/** 启动任务 */
export function startOtaTask(taskId: string) {
  return request.post(`/v1/ota/tasks/${taskId}/start`)
}

/** 取消任务 */
export function cancelOtaTask(taskId: string) {
  return request.post(`/v1/ota/tasks/${taskId}/cancel`)
}

/** 暂停任务 */
export function pauseOtaTask(taskId: string) {
  return request.post(`/v1/ota/tasks/${taskId}/pause`)
}

/** 恢复任务 */
export function resumeOtaTask(taskId: string) {
  return request.post(`/v1/ota/tasks/${taskId}/resume`)
}
